prov/gni/src/gnix_vc.c

```diff
@@ -69,6 +70,367 @@ static int __gnix_vc_conn_ack_comp_fn(void *data);
+ * ep is non-null;
+ * vc_ptr is non-null.
+ *
+ * TODO: Only lock the vc_lock when looking up the VC.
+ */
+static inline int __gnix_vc_lookup_vc(struct gnix_fid_ep *ep, void *dest_addr,
+				      bool is_dest_addr_gnix_addr,
+				      struct gnix_av_addr_entry **av_entry,
+				      gnix_ht_key_t **key,
+				      fi_addr_t *table_dest_addr,
+				      struct gnix_address *ht_dest_addr,
+				      struct gnix_vc **vc_ptr)
+{
+	struct gnix_fid_av *av = ep->av;
+	int av_type = av->type, ret = FI_SUCCESS;
+	gnix_ht_key_t *key_ptr = *key;
+
+	/*
+	 * Ensure that vc_ptr is NULL before vc lookup as
+	 * functions rely on vc_ptr being NULL to determine
+	 * whether the vc was found during the lookup.
+	 */
+	*vc_ptr = NULL;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+
+	if (likely((av_type == FI_AV_TABLE))) {
+		/* Check to see if the user passed in a gnix address */
+		if (is_dest_addr_gnix_addr) {
+			ret = _gnix_table_reverse_lookup(
+				av, *((struct gnix_address *) dest_addr),
+				table_dest_addr);
+
+                        GNIX_INFO(FI_LOG_EP_CTRL, "gnix_addr: %llx\n",
+                                  *((struct gnix_address *)dest_addr));
+                        if (ret != FI_SUCCESS) {
+				GNIX_WARN(FI_LOG_EP_CTRL,
+					  "_gnix_table_reverse_lookup returned %s\n",
+					  fi_strerror(-ret));
+				return ret;
+			}
+		} else {
+			*table_dest_addr = *((fi_addr_t *) dest_addr);
+
+                        GNIX_INFO(FI_LOG_EP_CTRL, "fi_addr_t: %llx\n",
+                                  *table_dest_addr);
+                }
+
+		ret = _gnix_table_lookup(av, *table_dest_addr, av_entry);
+		assert(*av_entry);
+
+		if (ret != FI_SUCCESS) {
+			GNIX_WARN(FI_LOG_EP_CTRL,
+				  "_gnix_table_lookup returned %s\n",
+				  fi_strerror(-ret));
+			return ret;
+		}
+		/* fastlock_acquire(&ep->vc_lock); */
+                ret = _gnix_vec_at(ep->vc_table, (void **)vc_ptr,
+                                   *table_dest_addr);
+                /* fastlock_release(&ep->vc_lock); */
+                GNIX_DEBUG(FI_LOG_EP_CTRL, "Found vc %p at vector index %llu\n",
+                           *vc_ptr, *table_dest_addr);
+
+                /* Ignore FI_ECANCELED as it's returned when looking up a entry
+                 * that is empty */
+                if (ret != -FI_ECANCELED && ret != FI_SUCCESS) {
+			GNIX_WARN(FI_LOG_EP_CTRL, "_gnix_vec_at returned %s\n",
+				  fi_strerror(-ret));
+                }
+
+		return ret;
+	} else if (av_type == FI_AV_MAP) {
+		if (key_ptr == NULL) {
+			/* Check to see if the user passed in a gnix address */
+			if (is_dest_addr_gnix_addr) {
+				*ht_dest_addr = *((struct gnix_address *)
+						  dest_addr);
+
+                                GNIX_INFO(FI_LOG_EP_CTRL, "gnix_addr: 0x%llx\n",
+					  *ht_dest_addr);
+
+				key_ptr = malloc(sizeof(gnix_ht_key_t));
+				assert(key_ptr);
+
+                                memcpy(key_ptr, ht_dest_addr,
+                                       sizeof(gnix_ht_key_t));
```

Comment:
I don't see why you would need to malloc any memory here. The lookup doesn't require any memory from the user of the hashtable.